keps/sig-instrumentation/3077-contextual-logging/README.md

```diff
@@ -0,0 +1,1000 @@
+I1026 16:21:00.461673  801139 binder.go:842] NominatedPods/Filter/VolumeBinding: "No matching volumes for PVC on node" pod="default/my-csi-app-inline-volume" node="127.0.0.1" default/my-csi-app-inline-volume-my-csi-volume="127.0.0.1"
+I1026 16:21:00.461724  801139 binder.go:971] NominatedPods/Filter/VolumeBinding: "Node has no accessible CSIStorageCapacity with enough capacity" pod="default/my-csi-app-inline-volume" node="127.0.0.1" pvc="default/my-csi-app-inline-volume-my-csi-volume" pvcSize=549755813888000 sc="csi-hostpath-fast"
+I1026 16:21:00.461817  801139 preemption.go:195] "Preemption will not help schedule pod on any node" pod="default/my-csi-app-inline-volume"
+I1026 16:21:00.461886  801139 scheduler.go:464] "Status after running PostFilter plugins for pod" pod="default/my-csi-app-inline-volume" status=&{code:2 reasons:[0/1 nodes are available: 1 Preemption is not helpful for scheduling.] err:<nil> failedPlugin:}
+I1026 16:21:00.461918  801139 factory.go:209] "Unable to schedule pod; no fit; waiting" pod="default/my-csi-app-inline-volume" err="0/1 nodes are available: 1 node(s) did not have enough free storage."
+```
+
+### Test Plan
+
+The new code will be covered by unit tests that execute as part of `pull-kubernetes-unit`.
+
+Converted components will be tested by exercising them with JSON output at high
+log levels to emit as many log messages as possible. Analysis of those logs
+will detect duplicate keys that might occur when a caller uses `WithValues` and
+a callee adds the same value in a log message. Static code analysis cannot
+detect this, or at least not easily.
+
+What it can check is that the individual log calls pass valid key/value pairs
+(strings as keys, always a matching value for each key, no duplicate keys).
+
+It can also detect usage of klog in code that should only use contextual
+logging.
+
+### Graduation Criteria
+
+#### Alpha
+
+- Common utility code available (logcheck with the additional checks,
+  `k8s.io/klogr` v0.1.0)
+- Documentation for developers available at https://github.com/kubernetes/community
+- At least kube-scheduler framework and some plugins converted
+- Initial e2e tests completed and enabled in Prow
+
+#### Beta
+
+- All of kube-scheduler (in-tree) and CSI external-provisioner (out-of-tree) converted
+- Users of klog notified through a Kubernetes blog post and an email to
+  dev@kubernetes.io that a logger must be set with k8s.io/klogr.
+- Issues filed against kubernetes-sigs projects with a request to set a logger
+  with k8s.io/klogr. These issues will have a link to the enhancement issue
+  (for cross-referencing) and a link to the developer documentation.
+
+#### GA
+
+- All code in kubernetes/kubernetes converted to contextual logging,
+  no dependency on klog anymore
+- All code under kubernetes-sigs sets a logger with k8s.io/klogr
+  and therefore is ready for the next k/k release.
+- Gather feedback from developers and surveys
```

Comment:
Gathering feedback should be rather done in Beta stage, as there should not be big changes before GA. For GA it we should rather have a point `User feedback is addressed`

Reply:
Okay.